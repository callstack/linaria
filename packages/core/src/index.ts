export { default as css } from './css';
export { default as cx } from './cx';
export type { CSSProperties } from './CSSProperties';
export type { LinariaClassName } from './cx';
