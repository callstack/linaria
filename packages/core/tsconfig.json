{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {},
    "rootDir": "src/",
    "types": ["node"]
  },
  "references": [{ "path": "../logger" }, { "path": "../utils" }]
}
