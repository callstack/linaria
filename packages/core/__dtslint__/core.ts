/* eslint-disable @typescript-eslint/no-unused-vars */
import { css, cx } from '../src';

// $ExpectType LinariaClassName
const class1 = css``;

const activeClass = css``;

// $ExpectType string
const combined1 = cx(class1, 'active');

// $ExpectType LinariaClassName
const combined2 = cx(class1, activeClass);
