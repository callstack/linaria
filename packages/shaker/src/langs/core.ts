import { types as t } from '@babel/core';
import type {
  AssignmentExpression,
  Block,
  CallExpression,
  Directive,
  ExpressionStatement,
  ForInStatement,
  ForStatement,
  Function,
  Identifier,
  IfStatement,
  MemberExpression,
  Node,
  ObjectExpression,
  SequenceExpression,
  SwitchCase,
  SwitchStatement,
  Terminatorless,
  TryStatement,
  VariableDeclaration,
  VariableDeclarator,
  WhileStatement,
} from '@babel/types';

import { peek } from '@linaria/babel-preset';
import type { IdentifierHandlers, Visitors } from '../types';
import GraphBuilderState from '../GraphBuilderState';
import ScopeManager from '../scope';
import DepsGraph from '../DepsGraph';

function isIdentifier(
  node: Node,
  name?: string | string[]
): node is Identifier {
  return (
    t.isIdentifier(node) &&
    (name === undefined ||
      (Array.isArray(name) ? name.includes(node.name) : node.name === name))
  );
}

type SideEffect = [
  {
    callee?: (child: CallExpression['callee']) => boolean;
    arguments?: (child: CallExpression['arguments']) => boolean;
  },
  (node: CallExpression, state: GraphBuilderState) => void
];

const sideEffects: SideEffect[] = [
  [
    // if the first argument of forEach is required, mark forEach as required
    {
      callee: (node) =>
        t.isMemberExpression(node) &&
        t.isIdentifier(node.property) &&
        node.property.name === 'forEach',
    },
    (node, state) => state.graph.addEdge(node.arguments[0], node),
  ],
];

function getCallee(node: CallExpression): Node {
  if (
    t.isSequenceExpression(node.callee) &&
    node.callee.expressions.length === 2
  ) {
    const [first, second] = node.callee.expressions;
    if (t.isNumericLiteral(first) && first.value === 0) {
      return second;
    }
  }

  return node.callee;
}

function findWildcardReexportStatement(
  node: t.CallExpression,
  identifierName: string,
  graph: DepsGraph
): t.Statement | null {
  if (!t.isIdentifier(node.callee) || node.callee.name !== 'require')
    return null;

  const declarator = graph.getParent(node);
  if (!t.isVariableDeclarator(declarator)) return null;

  const declaration = graph.getParent(declarator);
  if (!t.isVariableDeclaration(declaration)) return null;

  const program = graph.getParent(declaration);
  if (!t.isProgram(program)) return null;

  // Our node is a correct export
  // Let's check that we have something that looks like transpiled re-export
  return (
    program.body.find((statement) => {
      /*
       * We are looking for `Object.keys(_bar).forEach(…)`
       */

      if (!t.isExpressionStatement(statement)) return false;

      const expression = statement.expression;
      if (!t.isCallExpression(expression)) return false;

      const callee = expression.callee;
      if (!t.isMemberExpression(callee)) return false;

      const { object, property } = callee;

      if (!isIdentifier(property, 'forEach')) return false;

      if (!t.isCallExpression(object)) return false;

      // `object` should be `Object.keys`
      if (
        !t.isMemberExpression(object.callee) ||
        !isIdentifier(object.callee.object, 'Object') ||
        !isIdentifier(object.callee.property, 'keys')
      )
        return false;

      //
      const [argument] = object.arguments;
      return isIdentifier(argument, identifierName);
    }) ?? null
  );
}

/*
 * Returns nodes which are implicitly affected by specified node
 */
function getAffectedNodes(node: Node, state: GraphBuilderState): Node[] {
  // FIXME: this method should be generalized
  const callee = t.isCallExpression(node) ? getCallee(node) : null;
  if (
    t.isCallExpression(node) &&
    t.isMemberExpression(callee) &&
    isIdentifier(callee.object, 'Object') &&
    isIdentifier(callee.property, [
      'assign',
      'defineProperty',
      'defineProperties',
      'freeze',
      'observe',
    ])
  ) {
    const [obj, property] = node.arguments;
    if (!t.isIdentifier(obj)) {
      return [];
    }

    if (
      state.scope.getDeclaration(obj) !== ScopeManager.globalExportsIdentifier
    ) {
      return [node.arguments[0]];
    }

    if (t.isStringLiteral(property)) {
      if (property.value === '__esModule') {
        return [node.arguments[0]];
      }

      state.graph.addExport(property.value, node);
    }
  }

  return [];
}

export const visitors: Visitors = {
  /*
   * FunctionDeclaration | FunctionExpression | ObjectMethod | ArrowFunctionExpression | ClassMethod | ClassPrivateMethod;
   * Functions can be either a statement or an expression.
   * That's why we need to disable default dependency resolving strategy for expressions by passing `ignoreDeps` flag.
   * Every function must have a body. Without a body, it becomes invalid.
   * In general, a body depends on parameters of a function.
   * In real life, some of the parameters can be omitted, but it's not trivial to implement that type of tree shaking.
   */
  Function(this: GraphBuilderState, node: Function) {
    const unsubscribe = this.onVisit((descendant) =>
      this.graph.addEdge(node, descendant)
    );
    this.baseVisit(node, true); // ignoreDeps=true prevents default dependency resolving
    unsubscribe();

    this.graph.addEdge(node, node.body);

    node.params.forEach((param) => this.graph.addEdge(node.body, param));
    if (
      t.isFunctionExpression(node) &&
      node.id !== null &&
      node.id !== undefined
    ) {
      // keep function name in expressions like `const a = function a();`
      this.graph.addEdge(node, node.id);
    }
  },

  /*
   * ExpressionStatement
   */
  ExpressionStatement(this: GraphBuilderState, node: ExpressionStatement) {
    this.baseVisit(node);

    this.graph.addEdge(node, node.expression);
    this.graph.addEdge(node.expression, node);
  },

  /*
   * BlockStatement | Program
   * The same situation as in ExpressionStatement: if one of the expressions is required, the block itself is also required.
   * Whereas a block doesn't depend on its children.
   * Example:
   * 1. let c;
   * 2. { // BlockStatement begin
   * 3.   let a = 1;
   * 4.   let b = 2;
   * 5.   a++;
   * 6.   a = c;
   * 7. } // BlockStatement end
   *
   * If we want to evaluate the value of `c`, we need to evaluate lines 1, 3, 5 and 6,
   * but we don't need line 4, even though it's a child of the block.
   */
  Block(this: GraphBuilderState, node: Block) {
    this.baseVisit(node);

    if (t.isProgram(node)) {
      const exportsDeclaration = this.scope.getDeclaration('global:exports')!;
      this.graph.addEdge(node, exportsDeclaration);
      node.directives.forEach((directive) =>
        this.graph.addEdge(node, directive)
      );
    }
  },

  Directive(this: GraphBuilderState, node: Directive) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.value);
  },

  /*
   * TryStatement
   * try { /* block *\/ } catch() {/* handler *\/} finalize {/* finalizer *\/}
   * `handler` and `finalizer` do not make sense without `block`
   * `block` depends on the whole node.
   */
  TryStatement(this: GraphBuilderState, node: TryStatement) {
    this.baseVisit(node);
    [node.handler, node.finalizer].forEach((statement) => {
      if (statement) {
        this.graph.addEdge(node.block, statement);
        this.graph.addEdge(statement, node.block);
      }
    });
  },

  IfStatement(this: GraphBuilderState, node: IfStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.consequent);
    this.graph.addEdge(node, node.test);
  },

  /*
   * WhileStatement
   * Pretty simple behaviour here:
   * • if body is required, the statement is required
   * • if the statement is required, the condition is also required.
   */
  WhileStatement(this: GraphBuilderState, node: WhileStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.test);
  },

  SwitchCase(this: GraphBuilderState, node: SwitchCase) {
    this.baseVisit(node);
    node.consequent.forEach((statement) => this.graph.addEdge(statement, node));
    if (node.test) {
      this.graph.addEdge(node, node.test);
    }
  },

  SwitchStatement(this: GraphBuilderState, node: SwitchStatement) {
    this.baseVisit(node);
    node.cases.forEach((c) => this.graph.addEdge(c, node));
    this.graph.addEdge(node, node.discriminant);
  },

  ForStatement(this: GraphBuilderState, node: ForStatement) {
    this.baseVisit(node);

    [node.init, node.test, node.update, node.body].forEach((child) => {
      if (child) {
        this.graph.addEdge(node, child);
      }
    });
  },

  /*
   * ForInStatement
   * for (const k in o) { body }
   */
  ForInStatement(this: GraphBuilderState, node: ForInStatement) {
    this.baseVisit(node);

    if (node.body) {
      this.graph.addEdge(node, node.body);
      this.graph.addEdge(node.body, node.left);
    }

    this.graph.addEdge(node.left, node.right);
  },

  /*
   * BreakStatement | ContinueStatement | ReturnStatement | ThrowStatement | YieldExpression | AwaitExpression
   * All these nodes are required to evaluate the value of a function in which they are defined.
   * Also, the value of these nodes depends on the argument if it is presented.
   */
  Terminatorless(this: GraphBuilderState, node: Terminatorless) {
    this.baseVisit(node);

    if (
      !(t.isBreakStatement(node) || t.isContinueStatement(node)) &&
      node.argument
    ) {
      this.graph.addEdge(node, node.argument);
    }

    const closestFunctionNode = peek(this.fnStack);
    this.graph.addEdge(closestFunctionNode, node);
  },

  /*
   * ObjectExpression
   * Objects are… complicated. Especially because similarly looking code can be either an expression or a pattern.
   * In this case we work with an expression like:
   * const obj = {
   *   method() {}, // ObjectMethod
   *   property: "value", // ObjectProperty
   *   ...rest, // SpreadElement
   * }
   */
  ObjectExpression(this: GraphBuilderState, node: ObjectExpression) {
    this.context.push('expression');
    this.baseVisit(node);
    node.properties.forEach((prop) => {
      this.graph.addEdge(node, prop);
      if (t.isObjectMethod(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.body);
      } else if (t.isObjectProperty(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.value);
      } else if (t.isSpreadElement(prop)) {
        this.graph.addEdge(prop, prop.argument);
      }
    });
    this.context.pop();
  },

  /*
   * MemberExpression
   * It's about a simple expression like `obj.foo` or `obj['foo']`.
   * In addition to default behaviour (an expression depends on all its children),
   * we add a backward dependency from an object to a node for processing member
   * expressions in assignments.
   *
   * Example:
   * let obj = { a: 1 };
   * obj.b = 2;
   *
   * If we try to evaluate `obj` without backward dependency,
   * `obj.b = 2` will be cut and we will get just `{ a: 1 }`.
   */
  MemberExpression(this: GraphBuilderState, node: MemberExpression) {
    this.baseVisit(node);

    if (
      isIdentifier(node.object, 'exports') &&
      this.scope.getDeclaration(node.object) ===
        ScopeManager.globalExportsIdentifier
    ) {
      // We treat `exports.something` and `exports['something']` as identifiers in the global scope
      this.graph.addEdge(node, node.object);
      this.graph.addEdge(node, node.property);

      const isLVal = peek(this.context) === 'lval';
      if (isLVal) {
        this.scope.declare(node, false);
      } else {
        const declaration = this.scope.addReference(node);
        this.graph.addEdge(node, declaration);
      }

      return;
    }

    if (t.isIdentifier(node.object) && t.isIdentifier(node.property)) {
      // It's simple `foo.bar` expression. Is it a usage of a required library?
      const declaration = this.scope.getDeclaration(node.object);
      if (
        t.isIdentifier(declaration) &&
        this.graph.importAliases.has(declaration) &&
        !node.computed
      ) {
        // It is. We can remember what exactly we use from it.
        const source = this.graph.importAliases.get(declaration)!;
        this.graph.imports.get(source)!.push(node.property);
      }
    }
  },

  /*
   * AssignmentExpression
   * `a = b`, `{ ...rest } = obj`, `obj.a = 3`, etc.
   * It's not a declaration, it's just an assignment, but it affects
   * the value of declared variable if the variable it mentioned in the left part.
   * So, we apply some context-magic here in order to catch reference of variables in the left part.
   * We switch the context to `lval` and continue traversing through the left branch.
   * If we then meet some identifier, we mark it as a dependency of its declaration.
   */
  AssignmentExpression(this: GraphBuilderState, node: AssignmentExpression) {
    this.context.push('lval');
    this.visit<AssignmentExpression['left'], AssignmentExpression>(
      node.left,
      node,
      'left'
    );
    this.context.pop();

    this.visit(node.right, node, 'right');

    // The value of an expression depends on the left part.
    this.graph.addEdge(node, node.left);

    // The left part of an assignment depends on the right part.
    this.graph.addEdge(node.left, node.right);
  },

  /*
   * VariableDeclarator
   * It would be pretty simple if it weren't used to declare variables from other modules.
   */
  VariableDeclarator(this: GraphBuilderState, node: VariableDeclarator) {
    /*
     * declared is used for detecting external dependencies in cases like
     * const { a, b, c } = require('module');
     *
     * We are remembering all declared variables in order to use it later in CallExpression visitor
     */
    const declared: Array<[Identifier, Identifier | null]> = [];
    this.meta.set('declared', declared);
    const unregister = this.scope.addDeclareHandler((identifier, from) =>
      declared.push([identifier, from])
    );
    this.baseVisit(node);
    this.meta.delete('declared');
    unregister();

    if (node.init) {
      // If there is an initialization part, the identifier depends on it.
      this.graph.addEdge(node.id, node.init);
    }

    // If a statement is required itself, an id is also required
    this.graph.addEdge(node, node.id);
  },

  /*
   * VariableDeclaration
   * It's just a wrapper for group of VariableDeclarator.
   * If one of the declarators is required, the wrapper itself is also required.
   */
  VariableDeclaration(this: GraphBuilderState, node: VariableDeclaration) {
    this.meta.set('kind-of-declaration', node.kind);
    this.baseVisit(node);
    node.declarations.forEach((declaration) =>
      this.graph.addEdge(declaration, node)
    );
    this.meta.delete('kind-of-declaration');
  },

  /*
   * CallExpression
   * Do you remember that we have already mentioned it in VariableDeclarator?
   * It is a simple expression with default behaviour unless it is a `require`.
   *
   * Another tricky use case here is functions with side effects (e.g. `Object.defineProperty`).
   */
  CallExpression(
    this: GraphBuilderState,
    node: CallExpression,
    parent: Node | null
  ) {
    this.baseVisit(node);

    if (t.isIdentifier(node.callee) && node.callee.name === 'require') {
      // It looks like a module import …
      const scopeId = this.scope.whereIsDeclared(node.callee);
      if (scopeId && scopeId !== 'global') {
        // … but it is just a user defined function
        return;
      }

      const [firstArg] = node.arguments;
      if (!t.isStringLiteral(firstArg)) {
        // dynamic import? Maybe someday we can do something about it
        return;
      }

      const { value: source } = firstArg;
      const declared = this.meta.get('declared') as Array<
        [Identifier, Identifier | null]
      >;
      if (!declared) {
        // This is a standalone `require`
        return;
      }

      // Define all declared variables as external dependencies.
      declared.forEach(([local, _imported]) =>
        // FIXME: var slugify = require('../slugify').default;
        {
          if (!this.graph.imports.has(source)) {
            this.graph.imports.set(source, []);
          }

          if (
            parent &&
            t.isMemberExpression(parent) &&
            t.isIdentifier(parent.property)
          ) {
            // An imported function is specified right here.
            // eg. require('../slugify').default
            this.graph.imports.get(source)!.push(parent.property);
          } else {
            if (
              t.isCallExpression(parent) &&
              t.isIdentifier(parent.callee) &&
              typeof parent.callee.name === 'string'
            ) {
              if (parent.callee.name.startsWith('_interopRequireDefault')) {
                this.graph.importTypes.set(source, 'default');
              } else if (
                parent.callee.name.startsWith('_interopRequireWildcard')
              ) {
                this.graph.importTypes.set(source, 'wildcard');
              } else {
                // What I've missed?
              }
            }

            // Do we know the type of import?
            if (!this.graph.importTypes.has(source)) {
              // Is it a wildcard reexport? Let's check.
              const statement = findWildcardReexportStatement(
                node,
                local.name,
                this.graph
              );
              if (statement) {
                this.graph.addEdge(local, statement);
                this.graph.reexports.push(local);
                this.graph.importTypes.set(source, 'reexport');
              }
            }

            // The whole namespace was imported. We will know later, what exactly we need.
            // eg. const slugify = require('../slugify');
            this.graph.importAliases.set(local, source);
          }
        }
      );

      return;
    }

    sideEffects.forEach(([conditions, callback]) => {
      if (
        (conditions.callee && !conditions.callee(node.callee)) ||
        (conditions.arguments && !conditions.arguments(node.arguments))
      ) {
        return;
      }

      return callback(node, this);
    });

    getAffectedNodes(node, this).forEach((affectedNode) => {
      this.graph.addEdge(affectedNode, node);
      if (t.isIdentifier(affectedNode)) {
        this.graph.addEdge(
          this.scope.getDeclaration(affectedNode)!,
          affectedNode
        );
      }
    });
  },

  /*
   * SequenceExpression
   * It is a special case of expression in which the value of the whole
   * expression depends only on the last subexpression in the list.
   * The rest of the subexpressions can be omitted if they don't have dependent nodes.
   *
   * Example:
   * const a = (1, 2, b = 3, 4, b + 2); // `a` will be equal 5
   */
  SequenceExpression(this: GraphBuilderState, node: SequenceExpression) {
    // Sequence value depends on only last expression in the list
    this.baseVisit(node, true);
    if (node.expressions.length > 0) {
      this.graph.addEdge(node, node.expressions[node.expressions.length - 1]);
    }
  },
};

export const identifierHandlers: IdentifierHandlers = {
  declare: [
    ['CatchClause', 'param'],
    ['Function', 'params'],
    ['FunctionExpression', 'id'],
    ['RestElement', 'argument'],
    ['ThrowStatement', 'argument'],
    ['VariableDeclarator', 'id'],
  ],
  keep: [['ObjectProperty', 'key']],
  refer: [
    ['ArrayExpression', 'elements'],
    ['AssignmentExpression', 'left', 'right'],
    ['BinaryExpression', 'left', 'right'],
    ['CallExpression', 'arguments', 'callee'],
    ['ConditionalExpression', 'test', 'consequent', 'alternate'],
    ['ForInStatement', 'right'],
    ['Function', 'body'],
    ['IfStatement', 'test'],
    ['LogicalExpression', 'left', 'right'],
    ['NewExpression', 'arguments', 'callee'],
    ['ObjectProperty', 'value'],
    ['ReturnStatement', 'argument'],
    ['SequenceExpression', 'expressions'],
    ['SwitchStatement', 'discriminant'],
    ['UnaryExpression', 'argument'],
    ['UpdateExpression', 'argument'],
    ['VariableDeclarator', 'init'],
  ],
};
