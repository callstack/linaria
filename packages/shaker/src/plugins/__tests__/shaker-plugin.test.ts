import { join } from 'path';

import type { PluginItem } from '@babel/core';
import { transformSync } from '@babel/core';
import dedent from 'dedent';

import shakerPlugin, { hasShakerMetadata } from '../shaker-plugin';

type Extension = 'js' | 'ts' | 'jsx' | 'tsx';

const getPresets = (extension: Extension) => {
  const presets: PluginItem[] = [];
  if (extension === 'ts' || extension === 'tsx') {
    presets.push(require.resolve('@babel/preset-typescript'));
  }

  if (extension === 'jsx' || extension === 'tsx') {
    presets.push(require.resolve('@babel/preset-react'));
  }

  return presets;
};

const keep =
  (only: string[], extension: Extension = 'js') =>
  (code: TemplateStringsArray) => {
    const presets = getPresets(extension);
    const filename = join(__dirname, `source.${extension}`);
    const formattedCode = dedent(code);

    const transformed = transformSync(formattedCode, {
      babelrc: false,
      configFile: false,
      filename,
      presets,
      plugins: [
        [
          shakerPlugin,
          {
            onlyExports: only,
          },
        ],
      ],
    });

    if (
      !transformed ||
      !transformed.code ||
      !hasShakerMetadata(transformed.metadata)
    ) {
      throw new Error(`${filename} has no shaker metadata`);
    }

    return {
      code: transformed.code,
      metadata: transformed.metadata.__linariaShaker,
    };
  };

describe('shaker', () => {
  it('should remove unused export', () => {
    const { code, metadata } = keep(['a'])`
      export const a = 1;
      export const b = 2;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should remove related code', () => {
    const { code, metadata } = keep(['a'])`
      const a = 1;
      const b = 2;
      export { a, b };
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should process array patterns', () => {
    const { code, metadata } = keep(['c'])`
      const [a, b, c] = array;
      const [,,, d, e] = array;

      export { a, b, c, d, e};
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should process object patterns', () => {
    const { code, metadata } = keep(['b1'])`
      const { a: a1, b: b1, c: c1 } = obj;
      const { d: d1, e: e1 } = obj;

      export { a1, b1, c1, d1, e1 };
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should process transpiled enums', () => {
    const { code, metadata } = keep(['defaultValue'])`
      exports.Kind = void 0;

      var Kind;
      (function (Kind) {
        Kind["DEFAULT"] = "default";
        Kind["TRANSPARENT"] = "transparent";
        Kind["RED"] = "red";
        Kind["BLACK"] = "black";
      })(Kind = exports.Kind || (exports.Kind = {}));

      export const defaultValue = Kind.DEFAULT;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should keep referenced exports', () => {
    const { code, metadata } = keep(['defaultValue'])`
      exports.foo = 10;

      exports.defaultValue = Math.random() * exports.foo;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should throw out unused referenced exports', () => {
    const { code, metadata } = keep(['defaultValue'])`
      exports.foo = 10;

      exports.bar = Math.random() * exports.foo;

      exports.defaultValue = 20;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should remove all references of `a`', () => {
    const { code, metadata } = keep(['b'])`
      let a = 1;

      a = 2;

      exports.a = a;
      exports.b = 10;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should respect implicit references', () => {
    const { code, metadata } = keep(['a'])`
      let _a;
      exports.a = _a = {};
      exports.b = _a;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should keep assigment even if export is marked for removing', () => {
    const { code, metadata } = keep(['b'])`
      let _a;
      exports.a = _a = {};
      exports.b = _a;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should process identifiers in void expressions as references', () => {
    const { code, metadata } = keep(['b'])`
      let _;

      const a = void _;

      function b(b) {
        void _;
      }

      exports.a = a;
      exports.b = b;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should delete import', () => {
    const { code, metadata } = keep(['Alive'])`
      import { A, B } from "ABC";
      const AB = {
        A,
        B,
      };

      export const Alive = "";

      export default AB;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should handle object patterns in exports', () => {
    const { code } = keep(['Alive'])`
      import foo from "foo";

      export const { Alive, Dead } = foo();
    `;

    expect(code).toMatchInlineSnapshot(`
      "import foo from \\"foo\\";
      export const {
        Alive,
        Dead
      } = foo();"
    `);
  });

  it('avoids deleting non-default exports when importing default export of a module without an __esModule: true property', () => {
    /* without workaround, this will be transformed by shaker to:
      const n = require('n');
      const defaultExports = {
        createContext: n.createContext
      };
      exports.default = defaultExports;

      i.e, exports.createContext is deleted
    */
    const { code } = keep(['default'])`
    const n = require('n');
    const defaultExports = { createContext: n.createContext }
    Object.defineProperty(exports, "createContext", {
      enumerable: !0,
      get: function() {
          return n.createContext
      }
    })
    exports.default = defaultExports;
    `;

    /*
      this breaks babel's interopRequireDefault
      without shaker, interopRequireDefault(<the test module>)
      returns: {
        default: {
          default: { createContext: ... }
          createContext: ...
        }
      }
      The double-default is because the module does not define
      __esModule: true, so interopRequireDefault assumes that it needs to wrap the exports in { default: ... }, so that subsequent code can access createContext via `<test module>.default.createContext`.

      If shaker treats the createExport named export as unused (since it's not in onlyExports), it will delete it. interopRequireDefault(<shaker-processed test module>)
      returns: {
        default: {
          default: { createContext: ... }
        }
      }

      And `<test module>.default.createContext` will be undefined.

      Therefore, we assert that createContext is not deleted in this case.
    */
    expect(code).toMatchInlineSnapshot(`
      "const n = require('n');
      const defaultExports = {
        createContext: n.createContext
      };
      Object.defineProperty(exports, \\"createContext\\", {
        enumerable: !0,
        get: function () {
          return n.createContext;
        }
      });
      exports.default = defaultExports;"
    `);
  });

  it('deletes non-default exports when importing default export of a module with an __esModule: true property', () => {
    /* without workaround, this will be transformed by shaker to:
      const n = require('n');
      const defaultExports = {
        createContext: n.createContext
      };
      exports.default = defaultExports;

      i.e, exports.createContext is deleted
    */
    const { code } = keep(['default'])`
    const n = require('n');
    const defaultExports = { createContext: n.createContext }
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    Object.defineProperty(exports, "createContext", {
      enumerable: !0,
      get: function() {
          return n.createContext
      }
    })
    exports.default = defaultExports;
    `;

    expect(code).toMatchInlineSnapshot(`
      "const n = require('n');
      const defaultExports = {
        createContext: n.createContext
      };
      Object.defineProperty(exports, \\"__esModule\\", {
        value: true
      });
      Object.defineProperty(exports, \\"createContext\\", {
        enumerable: !0,
        get: function () {}
      });
      exports.default = defaultExports;"
    `);
  });

  it('should remove asset imports', () => {
    const { code, metadata } = keep(['a'])`
      import './asset.css';

      export const a = 1;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should keep side-effects from modules', () => {
    const { code, metadata } = keep(['a'])`
      import 'regenerator-runtime/runtime.js';

      export const a = 1;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(1);
  });

  it('should keep only side-effects', () => {
    const { code, metadata } = keep(['side-effect'])`
      import 'regenerator-runtime/runtime.js';

      export const a = 1;
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(1);
  });

  it('should handle __importDefault', () => {
    const { code, metadata } = keep(['Input'])`
      var __importDefault =
        (this && this.__importDefault) ||
        function (mod) {
          return mod && mod.__esModule ? mod : { default: mod };
        };
      Object.defineProperty(exports, '__esModule', { value: true });

      var Input_1 = require('./Input');
      Object.defineProperty(exports, 'Input', {
        enumerable: true,
        get: function () {
          return __importDefault(Input_1).default;
        },
      });
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(1);
    expect(metadata.imports.get('./Input')).toEqual(['default']);
  });

  it('should shake if __linariaPreval required but not exported', () => {
    const { code, metadata } = keep(['__linariaPreval', 'Input'])`
      import 'regenerator-runtime/runtime.js';

      export { Button } from "./Button";
      export { Input } from "./Input";
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(2);
    expect([...metadata.imports.keys()]).toEqual([
      'regenerator-runtime/runtime.js',
      './Input',
    ]);
  });

  it('should keep setBatch', () => {
    // A real-world example from react-redux
    const { code, metadata } = keep(['setBatch'])`
      function defaultNoopBatch(callback) {
        callback();
      }

      var batch = defaultNoopBatch;

      export var setBatch = function setBatch(newBatch) {
        return (batch = newBatch);
      };

      export var getBatch = function getBatch() {
        return batch;
      };
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });

  it('should process constant violations inside binding paths', () => {
    // Function `a` should be removed because it's only used in removed function `b`
    const { code, metadata } = keep(['c'])`
      function a(flag) { return (a = function(flag) { flag ? 1 : 2 }) }
      export function b() { return a(1) }
      export function c() {};
    `;

    expect(code).toMatchSnapshot();
    expect(metadata.imports.size).toBe(0);
  });
});
