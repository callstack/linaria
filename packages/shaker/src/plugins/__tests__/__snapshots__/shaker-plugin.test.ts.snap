// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shaker should delete import 1`] = `"export const Alive = \\"\\";"`;

exports[`shaker should handle __importDefault 1`] = `
"var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    default: mod
  };
};
Object.defineProperty(exports, '__esModule', {
  value: true
});
var Input_1 = require('./Input');
Object.defineProperty(exports, 'Input', {
  enumerable: true,
  get: function () {
    return __importDefault(Input_1).default;
  }
});"
`;

exports[`shaker should keep assigment even if export is marked for removing 1`] = `
"let _a;
exports.a = _a = {};
exports.b = _a;"
`;

exports[`shaker should keep only side-effects 1`] = `"import 'regenerator-runtime/runtime.js';"`;

exports[`shaker should keep referenced exports 1`] = `
"var _foo;
_foo = 10;
exports.defaultValue = Math.random() * _foo;
exports.foo = _foo;"
`;

exports[`shaker should keep setBatch 1`] = `
"function defaultNoopBatch(callback) {
  callback();
}
var batch = defaultNoopBatch;
export var setBatch = function setBatch(newBatch) {
  return batch = newBatch;
};"
`;

exports[`shaker should keep side-effects from modules 1`] = `
"import 'regenerator-runtime/runtime.js';
export const a = 1;"
`;

exports[`shaker should process array patterns 1`] = `
"const [,, c] = array;
export { c };"
`;

exports[`shaker should process constant violations inside binding paths 1`] = `
"export function c() {}
;"
`;

exports[`shaker should process identifiers in void expressions as references 1`] = `
"let _;
function b(b) {
  void _;
}
exports.b = b;"
`;

exports[`shaker should process object patterns 1`] = `
"const {
  b: b1
} = obj;
export { b1 };"
`;

exports[`shaker should process transpiled enums 1`] = `
"var _Kind;
exports.Kind = void 0;
var Kind;
(function (Kind) {
  Kind[\\"DEFAULT\\"] = \\"default\\";
  Kind[\\"TRANSPARENT\\"] = \\"transparent\\";
  Kind[\\"RED\\"] = \\"red\\";
  Kind[\\"BLACK\\"] = \\"black\\";
})(Kind = _Kind || (_Kind = {}));
export const defaultValue = Kind.DEFAULT;
exports.Kind = _Kind;"
`;

exports[`shaker should remove all references of \`a\` 1`] = `"exports.b = 10;"`;

exports[`shaker should remove asset imports 1`] = `"export const a = 1;"`;

exports[`shaker should remove related code 1`] = `
"const a = 1;
export { a };"
`;

exports[`shaker should remove unused export 1`] = `"export const a = 1;"`;

exports[`shaker should respect implicit references 1`] = `
"let _a;
exports.a = _a = {};
exports.b = _a;"
`;

exports[`shaker should shake if __linariaPreval required but not exported 1`] = `
"import 'regenerator-runtime/runtime.js';
export { Input } from \\"./Input\\";"
`;

exports[`shaker should throw out unused referenced exports 1`] = `"exports.defaultValue = 20;"`;
